### PR TITLE
feat: Add wrapper for defining columns in TTree in ROOT

### PR DESCRIPTION
**Pull request by laf070810**
Adds a wrapper for defining columns in TTree in [the ROOT software package](https://root.cern.ch/). The ntuple ROOT files in the test are from the test of the `hadd` wrapper. 

### QC

* [x] I confirm that:

For all wrappers added by this PR, 

* there is a test case which covers any introduced changes,
* `input:` and `output:` file paths in the resulting rule can be changed arbitrarily,
* either the wrapper can only use a single core, or the example rule contains a `threads: x` statement with `x` being a reasonable default,
* rule names in the test case are in [snake_case](https://en.wikipedia.org/wiki/Snake_case) and somehow tell what the rule is about or match the tools purpose or name (e.g., `map_reads` for a step that maps reads),
* all `environment.yaml` specifications follow [the respective best practices](https://stackoverflow.com/a/64594513/2352071),
* the `environment.yaml` pinning has been updated by running `snakedeploy pin-conda-envs environment.yaml` on a linux machine,
* wherever possible, command line arguments are inferred and set automatically (e.g. based on file extensions in `input:` or `output:`),
* all fields of the example rules in the `Snakefile`s and their entries are explained via comments (`input:`/`output:`/`params:` etc.),
* `stderr` and/or `stdout` are logged correctly (`log:`), depending on the wrapped tool,
* temporary files are either written to a unique hidden folder in the working directory, or (better) stored where the Python function `tempfile.gettempdir()` points to (see [here](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir); this also means that using any Python `tempfile` default behavior works),
* the `meta.yaml` contains a link to the documentation of the respective tool or command,
* `Snakefile`s pass the linting (`snakemake --lint`),
* `Snakefile`s are formatted with [snakefmt](https://github.com/snakemake/snakefmt),
* Python wrapper scripts are formatted with [black](https://black.readthedocs.io).
* Conda environments use a minimal amount of channels, in recommended ordering. E.g. for bioconda, use (conda-forge, bioconda, nodefaults, as conda-forge should have highest priority and defaults channels are usually not needed because most packages are in conda-forge nowadays).

### CODE REVIEW

File: phys/root/define_columns/environment.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+channels:
+  - conda-forge
+dependencies:
+  - ROOT>6.24
```
Comment (fgvieira) [inline phys/root/define_columns/environment.yaml:4]:
```suggestion
  - ROOT=6.30.4
```

File: phys/root/define_columns/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+authors:
+  - Anfeng Li
+input:
+  - ROOT files containing TTree to be processed.
+output:
+  - |
+    ROOT files with the newly defined columns saved.
+
+    NOTE: If "input" is a list of ROOT filenames, then "output" must have the same number of filenames, 
+    and one output file corresponds to one input file in order.
```
Comment (fgvieira) [inline phys/root/define_columns/meta.yaml:13]: Can you add the notes at the end?

File: phys/root/define_columns/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  - |
+    branches (optional): branches to be defined, 
+    specified in the format of [["branch_name1", "definition_expression1"], ["branch_name2", "definition_expression2"]]. 
+    If not specified, no branch will be defined, i.e. just save the TTree.
+  - |
+    redefine (optional): list of branch name to be redefined. 
+    If a branch has already been defined in the input TTree, its name must be in this redefine list.
+  - |
+    branches_to_save (optional): list of branch names to be saved. 
+    If not specified, all branches of the input TTree will be saved.
```
Comment (fgvieira) [inline phys/root/define_columns/meta.yaml:26]:
```suggestion
  - input_tree_name: name of the input TTree
  - output_tree_name: name of the output TTree
  - branches: branches to be defined, specified in the format of [["branch_name1", "definition_expression1"], ["branch_name2", "definition_expression2"]]. (optional)
    If not specified, no branch will be defined, i.e. just save the TTree.
  - redefine: list of branch name to be redefined. If a branch has already been defined in the input TTree, its name must be in this redefine list. (optional)
  - branches_to_save: list of branch names to be saved. If not specified, all branches of the input TTree will be saved. (optional)
```

File: phys/root/define_columns/wrapper.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ROOT.EnableImplicitMT(snakemake.threads)
+
+redefine_list = snakemake.params.get("redefine", [])
+input_tree_name = snakemake.params.get("input_tree_name", "")
+output_tree_name = snakemake.params.get("output_tree_name", "")
+branches = snakemake.params.get("branches", [])
+branches_to_save = snakemake.params.get("branches_to_save", None)
+
+for i in range(len(snakemake.input)):
+    df = ROOT.RDataFrame(input_tree_name, snakemake.input[i])
```
Comment (fgvieira) [inline phys/root/define_columns/wrapper.py:18]: If it is mandatory:
```suggestion
    df = ROOT.RDataFrame(snakemake.params.input_tree_name, snakemake.input[i])
```

File: phys/root/define_columns/wrapper.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        else:
+            df = df.Define(branch_name, branch_definition)
+    if branches_to_save is not None:
+        df.Snapshot(
+            output_tree_name,
+            snakemake.output[i],
+            branches_to_save,
+        )
+    else:
+        df.Snapshot(output_tree_name, snakemake.output[i])
```
Comment (fgvieira) [inline phys/root/define_columns/wrapper.py:31]: If it is mandatory:
```suggestion
            snakemake.params.output_tree_name,
            snakemake.output[i],
            branches_to_save,
        )
    else:
        df.Snapshot(snakemake.params.output_tree_name, snakemake.output[i])
```

File: phys/root/define_columns/meta.yaml (hunk 2 above)
Comment (laf070810) [reply phys/root/define_columns/meta.yaml]: Moved.

**Comment by laf070810** [comment]
The linting test should be ok now.

**Comment by laf070810** [comment]
> From what I can see, the wrapper takes several input files and each will generate an output file. If so, wouldn't make more sense to have the wrapper process a single file, and let `snakemake` take care of the several files?

Do you mean using wildcards to take care of several files? Yes, this should be the most common case, and this wrapper already supports single file processing. As I imagined, the codes for dealing with multiple files are added as some kind of "sugar", in case someone doesn't want to add a wildcard because of inconvenience or something.

**Comment by fgvieira** [comment]
Unless there is a specific need for such behavior, I think it is better to keep the wrappers as simple as possible and not try to do things that `snakemake` is more suitable to.

File: phys/root/define_columns/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  - single ROOT file containing the TTree to be processed.
+output:
+  - single ROOT file containing the TTree with the newly defined columns saved.
+params:
+  - input_tree_name: name of the input TTree
+  - output_tree_name: name of the output TTree
+  - branches: branches to be defined, specified in the format of [["branch_name1", "definition_expression1"], ["branch_name2", "definition_expression2"]]. (optional)
+      If not specified, no branch will be defined, i.e. just save the TTree.
+  - redefine: list of branch names to be redefined. If a branch has already been defined in the input TTree, its name must be in this redefine list. (optional)
+  - branches_to_save: list of branch names to be saved. If not specified, all branches of the input TTree will be saved. (optional)
```
Comment (fgvieira) [inline phys/root/define_columns/meta.yaml:16]:
```suggestion
input:
  - TTree ROOT file
output:
  - TTree ROOT file
params:
  - input_tree_name: name of the input TTree
  - output_tree_name: name of the output TTree
  - branches: branches to be defined, specified in the format of [["branch_name1", "definition_expression1"], ["branch_name2", "definition_expression2"]]. If not specified, no branch will be defined, i.e. just save the TTree.  (optional)
  - redefine: list of branch names to be redefined. It must be defined in `params.branch` also. (optional)
  - branches_to_save: list of branch names to be saved. If not specified, all branches of the input TTree will be saved. (optional)
```

**Review by fgvieira — APPROVED** [review]

**Pull request merged**